### ISSUE TITLE
Bug report: make prep leaves behind a bug-report file on certain platforms

### ISSUE THREAD

**Issue by lcn2** [ioccc-src/mkiocccentry#484]
On RHEL7.9 and Debian 10.13, the `make prep` command leaves behind a _bug-report.YYYYMMDD.HHMMSS.txt_ file.

Here is the  _bug-report.YYYYMMDD.HHMMSS.txt_ file, after `make prep` was performed on a RHEL7.9 based host:

[bug-report.20230123.134546.txt](https://github.com/ioccc-src/mkiocccentry/files/10483876/bug-report.20230123.134546.txt)

Here is the  _bug-report.YYYYMMDD.HHMMSS.txt_ file, after `make prep` was performed on a Debian 10.13 based host:

[bug-report.20230123.134539.txt](https://github.com/ioccc-src/mkiocccentry/files/10483891/bug-report.20230123.134539.txt)

Unless `make prep` fails, the _bug-report.YYYYMMDD.HHMMSS.txt_ file should be removed.

BTW: Although we cannot perform `make release` on those platforms (because their Bison/Flex are too far behind to allow `make parser-o` to be performed), the same removal of the _bug-report.YYYYMMDD.HHMMSS.txt_ file should be performed of `make release` succeeds.

**Comment by xexyl** [comment]
That's really strange. Yesterday I fixed a place where the `X_FLAG` was not set to empty and this solved that problem but maybe there's another path that is missed?

**Comment by xexyl** [comment]
Checking if it's triggered under CentOS. If it is and it's immediately obvious I'll look at fixing it but otherwise I'll have to do this tomorrow I'm afraid. I'll address it before the other things in that case.

**Comment by xexyl** [comment]
Oh I might know why. Checking the log you uploaded. One moment.

**Comment by xexyl** [comment]
I do indeed know why.

It's because if you look at the log you'll see some warnings were triggered. They're not issues but the fact there are warnings are there is why it still is there. But then the `-l` option makes it not seen.

Look at the end of the log for example:

```
One or more POSSIBLE issues detected:


    WARNING: checknr VERSION UNKNOWN
    WARNING: ctags VERSION UNKNOWN
    WARNING: gdate VERSION UNKNOWN
    WARNING: bison VERSION UNKNOWN
    WARNING: flex VERSION UNKNOWN
```

That's why it remains. So I'm not sure if this is a bug. Maybe it could be changed though but I think it's good to have it stay there esp for public overview.

It's why I do not recommend they run `make prep` but rather `make bug_report`.

Do you want it changed though? If so to what? I'll do it tomorrow in that case. Have to go now.

**Comment by xexyl** [comment]
> I do indeed know why.
> 
> It's because if you look at the log you'll see some warnings were triggered. They're not issues but the fact there are warnings are there is why it still is there. But then the `-l` option makes it not seen.
> 
> Look at the end of the log for example:
> 
> ```
> One or more POSSIBLE issues detected:
> 
> 
>     WARNING: checknr VERSION UNKNOWN
>     WARNING: ctags VERSION UNKNOWN
>     WARNING: gdate VERSION UNKNOWN
>     WARNING: bison VERSION UNKNOWN
>     WARNING: flex VERSION UNKNOWN
> ```
> 
> That's why it remains. So I'm not sure if this is a bug. Maybe it could be changed though but I think it's good to have it stay there esp for public overview.
> 
> It's why I do not recommend they run `make prep` but rather `make bug_report`.
> 
> Do you want it changed though? If so to what? I'll do it tomorrow in that case. Have to go now.

That being said some of them might be classified differently? For instance maybe checknr, ctags and gdate should be not a warning? Report it maybe but don't mark it as a warning?

Other tools and other tests might also be that way. What do you think?

I might be able to reply to comments here a bit later but if not then tomorrow.

**Comment by lcn2** [comment]
> It's why I do not recommend they run make prep

The `make prep` rule is in the section of the Makefile that is intended for repo maintainers, although other folks are free to use them if they understand the consequences. 🤓

**Comment by lcn2** [comment]
> That being said some of them might be classified differently? For instance maybe checknr, ctags and gdate should be not a warning? Report it maybe but don't mark it as a warning?
> 
> Other tools and other tests might also be that way. What do you think?
> 
> 

Good idea, perhaps they should simply be a "Notice" .. not quite something to warn someone about, just take note.  For example:

```
    Notice: checknr VERSION UNKNOWN
    Notice: ctags VERSION UNKNOWN
    Notice: gdate VERSION UNKNOWN
    Notice: bison VERSION UNKNOWN
    Notice: flex VERSION UNKNOWN
```

**Comment by xexyl** [comment]
The above are some nice ideas. I'll take a look at this and hopefully I can come up with something soon.

After that I hope to look at the man pages - for formatting and consistency checks - but it's possible I won't get the latter done in one day and almost certainly won't in one sitting.

**Comment by xexyl** [comment]
> > That being said some of them might be classified differently? For instance maybe checknr, ctags and gdate should be not a warning? Report it maybe but don't mark it as a warning?
> > Other tools and other tests might also be that way. What do you think?
> 
> Good idea, perhaps they should simply be a "Notice" .. not quite something to warn someone about, just take something to take note. For example:
> 
> ```
>     Notice: checknr VERSION UNKNOWN
>     Notice: ctags VERSION UNKNOWN
>     Notice: gdate VERSION UNKNOWN
>     Notice: bison VERSION UNKNOWN
>     Notice: flex VERSION UNKNOWN
> ```
> 

That's easy enough to do: simply prepend it to each warning message (I mean WARNING_SUMMARY). Though I will capitalise 'Notice' entirely.

> So `make prep` would still be able exit 0 in light of notice messages. In particular, `prep.sh` would be able to exit 0 even if "Notice:" messages were issued.
> 
I was wondering about that but I wonder how to go about it.

> Perhaps `prep.sh` should print something like:
> 
> ```
> All tests PASSED, 5 Notices issued.
> 

How to identify the number of notices? I don't know that it's possible to return another number other than an exit code from one script to another but it might be that `bug_report.sh` can be given a log file name to use and the `prep.sh` might also use the same format (file name format) so that then at the end of `prep.sh` it can use `grep` on it to determine the number of notices.

But then the fact that all tests passed is on the same line as the notices when the notice is specific to bug_report.sh could be confusing? But then again maybe not because it's a summary.

Perhaps even having it control the log file is good in another way because it can then tell the user the name of the bug report file (if any problems) so that they can investigate it if they need to.

> See test_ioccc/test_ioccc.log for more details.
> ```
> 
> No "Notice:" messages would be written to stdout. They would only be found in the `test_ioccc/test_ioccc.log` file.
> 
> To see them, one could just do:
> 
> ```shell
> grep Notice: test_ioccc/test_ioccc.log
> ```

Or else if the bug report file is still there (as it would be since the log file has to be kept for normal use of `bug_report.sh`) it can also refer them to that file?

> 
> And see such notice messages.
> 
> While we are on the subject of `prep.sh`, when it does fail, perhaps printing something like:
> 
> ```
> 3 tests failed, 5 Notices issued.
> 
> See test_ioccc/test_ioccc.log for more details."
> ```
> 
> Would be better.

That looks nicer indeed.

> 
> Nothing too fancy. Keep the output of `prep.sh` simple, compact and easy to read while leaving the details to the `test_ioccc/test_ioccc.log` file.
> 

Makes sense.

> Just have `prep.sh` end with printing either:
> 
> ```
> All tests PASSED, X Notices issued.
> 
> See test_ioccc/test_ioccc.log for more details.
> ```
> 
> Or:
> 
> ```
> Y tests failed, X Notices issued.
> 
> See test_ioccc/test_ioccc.log for more details.
> ```
> 
> And write the Warnings and Notices to the log file.

Of course.

**Comment by xexyl** [comment]
Hmm ... 

Another way would be to designate a special exit code for warnings in bug_report.sh. If the warning summary (which I renamed notice) is not empty at the end then it would exit that code unless the failure summary was not empty in which case it would exit the appropriate exit code.

This might be more problematic in some ways but in other ways it would simplify it.

Thinking about this. I'm making good progress on what I already did but I'm wondering about testing all the paths.

If the bug_report.sh had a special exit code for warnings the prep.sh could check for it without any problems and could still exit 0. It would also still decide the log file for bug report (already that's done). It has to have a separate function in prep.sh anyway.

The above ideas you and I both had could still work though just as well if I'm thinking this well enough. If a user runs the bug_report.sh manually they would see the warning anyway unless they specify `-l` but in that case they would not be warned and how else will they know about it (without checking the log file which if it exits 0 they might not think to) unless the exit code is non-zero?

Just some thoughts I'm thinking about ...

**Comment by xexyl** [comment]
> Hmm ...
> 
> Another way would be to designate a special exit code for warnings in bug_report.sh. If the warning summary (which I renamed notice) is not empty at the end then it would exit that code unless the failure summary was not empty in which case it would exit the appropriate exit code.
> 
> This might be more problematic in some ways but in other ways it would simplify it.
> 
> Thinking about this. I'm making good progress on what I already did but I'm wondering about testing all the paths.
> 
> If the bug_report.sh had a special exit code for warnings the prep.sh could check for it without any problems and could still exit 0. It would also still decide the log file for bug report (already that's done). It has to have a separate function in prep.sh anyway.
> 
> The above ideas you and I both had could still work though just as well if I'm thinking this well enough. If a user runs the bug_report.sh manually they would see the warning anyway unless they specify `-l` but in that case they would not be warned and how else will they know about it (without checking the log file which if it exits 0 they might not think to) unless the exit code is non-zero?
> 
> Just some thoughts I'm thinking about ...

Another way this would be useful: in the prep.sh script in the bug report action how would it know if it has a warning or not? It would have to grep through the log file. This could still be done but it would be better (maybe!) if it does this by an exit code that's not 0: if it is non-zero but not the special value (let's say: 142 or something to keep it way out of the way) then there was at least one error; if it's 142 (or whatever) a warning was triggered and `prep.sh` would now have an easier way to tell the user.

But since the `-l` option was specified they won't see it. If it's the special exit code the script prep.sh can then warn them about it.

Thinking about this and playing with this more ....

**Comment by xexyl** [comment]
Solved it! I have to clean up formatting now and then check with and without `-l` specified. If all goes as I think it should then I can make a commit and pull request.


But I might have to go afk for a bit before I can do that.

**Comment by xexyl** [comment]
Fixed in https://github.com/xexyl/mkiocccentry/commit/4bae627277eddb1e5ec68f75f10fbd7b16a23be0.

**Comment by xexyl** [comment]
Fixed in https://github.com/ioccc-src/mkiocccentry/pull/485/commits/4bae627277eddb1e5ec68f75f10fbd7b16a23be0.

**Comment by xexyl** [comment]
With commits https://github.com/ioccc-src/mkiocccentry/commit/4bae627277eddb1e5ec68f75f10fbd7b16a23be0 and https://github.com/ioccc-src/mkiocccentry/commit/fca9b905ae8f9db16dba49420c874dcda27ae53b can this issue be closed as resolved ?

**Comment by lcn2** [comment]
> With commits [4bae627](https://github.com/ioccc-src/mkiocccentry/commit/4bae627277eddb1e5ec68f75f10fbd7b16a23be0) and [fca9b90](https://github.com/ioccc-src/mkiocccentry/commit/fca9b905ae8f9db16dba49420c874dcda27ae53b) can this issue be closed as resolved ?

We are fixing things now ...

**Comment by xexyl** [comment]
> > With commits [4bae627](https://github.com/ioccc-src/mkiocccentry/commit/4bae627277eddb1e5ec68f75f10fbd7b16a23be0) and [fca9b90](https://github.com/ioccc-src/mkiocccentry/commit/fca9b905ae8f9db16dba49420c874dcda27ae53b) can this issue be closed as resolved ?
> 
> We are fixing things now ...

I'm curious about this but as I'll be leaving shortly I guess I'll find out tomorrow.

Hopefully it's just a clean up and not something I missed? I was quite tired this morning and later in the morning I discovered that I ate last night without even knowing it. I know because a container with cereal was thrown out. I don't know how much I ate but I ate some anyway and I have no recollection of it. :( 

Well anyway I'll check it out tomorrow. Shutting down momentarily. Can barely see.

Good night!

**Comment by lcn2** [comment]
Completed the fix of this bug report with commit 3f169c032629f076623a38491ad4b0d3a1fdbfae and commit 9b9c850d4f46ef0541aa0ba9efeef1fd36a14e18 and commit b781b73f4c7a7a9565ddd460fae1a4ecd48e5a08

And with those commits, this issue is now closed.

**Comment by xexyl** [comment]
> Completed the fix of this bug report with commit [3f169c0](https://github.com/ioccc-src/mkiocccentry/commit/3f169c032629f076623a38491ad4b0d3a1fdbfae) and commit [9b9c850](https://github.com/ioccc-src/mkiocccentry/commit/9b9c850d4f46ef0541aa0ba9efeef1fd36a14e18) and commit [b781b73](https://github.com/ioccc-src/mkiocccentry/commit/b781b73f4c7a7a9565ddd460fae1a4ecd48e5a08) and commit commit [1228a17](https://github.com/ioccc-src/mkiocccentry/commit/1228a1743499cb25db4607b7ab4b886476c7495b)
> 
> And with those commits, this issue is now closed.
> 
> ### UPDATE 0
> It would be a good idea, @xexyl to check our work in those commits above, just to be extra sure all is well.

Sure!

I am still at the point where I am unable to lie down but not quite awake enough to work on the repo. Since there is no chance of the former I hope that I will be more awake sooner than later.

**Comment by xexyl** [comment]
That being said I wonder about if it's a good idea to remove the log file if an optional tool is missing. It might be that in som cases this is okay but depending on the tool it could be helpful to help us identify a problem.

Now I believe this is only with the version checks at this point but is it possible that there are other places this could happen?

I think it's useful information to have that a tool is missing. After all why even check for it if we don't keep the report or note it to the user?

But that's all I can do now. Going to do something mindless as unfortunately that's all I am right now :(
